### PR TITLE
github util class added

### PR DESCRIPTION
**Pull request by rachejazz**
github engine was not used in github search module. Now it does

This PR solves issue #95

### CODE REVIEW

File: core/util/github.py [hunk 1]
```diff
@@ -26,75 +25,46 @@ def __init__(self, q, cookie, _type='Repositories', limit=1):
...
+		self.q = q
 		self.cookie = cookie
 		self.limit = limit
-		self._pages = ''
-		self.github = 'github.com'
-		self.type = _type
-		self.types = ['users', 'Repositories']
-
+		self._pages = 10
+		self._page = {"Github":[""]}
```
Comment (z3y50n) [inline core/util/github.py]: use single quotes for simple strings

File: core/util/github.py [hunk 2]
```diff
@@ -26,75 +25,46 @@ def __init__(self, q, cookie, _type='Repositories', limit=1):
...
+		self._links = ''
+		self.github_api = 'api.github.com'
+		self.types = ['users', 'repositories']
+		self.urls = [f"https://{self.github_api}/search/{self.types[0]}?q={self.q}&per_page={self.limit}",
+					f"https://{self.github_api}/search/{self.types[1]}?q={self.q}&per_page={self.limit}&page={self._pages}"
+					]
```
Comment (z3y50n) [inline core/util/github.py]: align them correctly to improve readability

File: core/util/github.py [hunk 3]
```diff
@@ -26,75 +25,46 @@ def __init__(self, q, cookie, _type='Repositories', limit=1):
...
-		max_attempt = len(urls)
-		for url in range(max_attempt):
-			self.framework.verbose(f"[GITHUB] Searching in {url} page...")
+		max_attempt = len(self.urls)
+		for page in range(max_attempt):
 			try:
-				req = self.framework.request(url=urls[url], headers={'Cookie': self.cookie}, allow_redirects=True)
+				self.framework.verbose(f"[GITHUB] Searching in {page} page...")
+				req = self.framework.request(url=self.urls[page], headers={'Cookie': self.cookie, 'Accept': 'application/vnd.github.v3.text-match+json'}, allow_redirects=True)
+				result = req.text
```
Comment (z3y50n) [inline core/util/github.py]: This is unused variable

File: core/util/github.py [hunk 4]
```diff
@@ -26,75 +25,46 @@ def __init__(self, q, cookie, _type='Repositories', limit=1):
...
+				result_json = req.json()
+				for results in result_json["items"]:
```
Comment (z3y50n) [inline core/util/github.py]: single quotes

File: core/util/github.py [hunk 5]
```diff
@@ -26,75 +25,46 @@ def __init__(self, q, cookie, _type='Repositories', limit=1):
...
+					if 'fork' not in results:
+						self._links = self._links + "".join(results['html_url']).replace("https://", " ")
```
Comment (z3y50n) [inline core/util/github.py]: single quotes

File: core/util/github.py [hunk 6]
```diff
@@ -26,75 +25,46 @@ def __init__(self, q, cookie, _type='Repositories', limit=1):
...
+					else:
+						self._page['Github'].append(results['html_url'])
+				repo = self._page['Github']
+				users = self._links.split(' ')
```
Comment (z3y50n) [inline core/util/github.py]: unused variables

File: core/util/github.py [hunk 7]
```diff
@@ -26,75 +25,46 @@ def __init__(self, q, cookie, _type='Repositories', limit=1):
...
 
 	@property
-	def all_emails(self):
-		parser = self.framework.page_parse(self._pages)
-		parser.pclean
-		emails = [("".join([chr(int(char[1:3], 16)) for char in email.split('&#')[1:]]))\
-					for email in parser.findall(r"(&#x\w+;)+")]
-		return list(set(emails))
+	def users(self):
+		return users
```
Comment (z3y50n) [inline core/util/github.py]: Not defined

File: core/util/github.py [hunk 8]
```diff
@@ -26,75 +25,46 @@ def __init__(self, q, cookie, _type='Repositories', limit=1):
...
-				final_output[i] = "<Withoud Description>"
-		return final_output
-
-	@property
-	def users(self):
-		parser = self.framework.page_parse(self._pages)
-		parser.pclean
-		links = [f"https://github.com/{x}" for x in parser.findall(r"Unfollow ([\w_\.]+)") if x != "this"]
-		return list(set(links))
+		return repo
```
Comment (z3y50n) [inline core/util/github.py]: Not defined

File: modules/search/github.py [hunk 9]
```diff
@@ -36,19 +36,23 @@
...
 		attr = engine(q, limit)
 	else:
 		attr = engine(q)
-
-	attr.run_crawl()
-	LINKS += attr.links
-	PAGES += attr.pages
+	if eng == 'github':
+		run = self.github(q)
+		run.run_crawl()
```
Comment (z3y50n) [inline modules/search/github.py:53]: It doesn't return anything and therefore I don't see any results.
Also you could try to figure a way to include `github` engine in the above 2 if statements and not run a separate `if`

**Comment by saeeddhqan** [comment]
Write the try/except/else statement more readable. Don't put all of the codes on the try block.

File: modules/search/github.py (hunk 9 above)
Comment (rachejazz) [reply modules/search/github.py:53]: I have tried various permutations. It couldn't be done. Do make the changes, I will follow it next time.

File: core/util/github.py (hunk 8 above)
Comment (rachejazz) [reply core/util/github.py]: Yes, my bad. I have added the changes. Thank you for pointing that out.

File: modules/search/github.py [hunk 10]
```diff
@@ -70,18 +79,22 @@ def module_api(self):
...
+		output['usernames'].append(user)
 
 	for link in self.reglib().filter(r"https://(www\.)?github\.com/[\w-]{1,39}/[\w\-\.]+", list(set(LINKS))):
 		repo = re.search(r"(https://(www\.)?github\.com/([\w-]{1,39})/[\w\-\.]+)", link)
 		if repo.group(3) not in bactery:
 			repo = repo.group(0)
 			if repo not in output['repositories']:
 				output['repositories'].append(repo)
-
+	for link in REPO:
```
Comment (saeeddhqan) [inline modules/search/github.py:96]: why for?

Comment (rachejazz) [reply modules/search/github.py:96]: REPO is list of repos coming from github api. It's not scraping from webpages but send repositories based on the keyword query directly. And yes,  no json formatting was needed. List of matched strings according to match-header

Comment (rachejazz) [reply modules/search/github.py:96]: any problem with the for loop? I saw usernames getting appended in the same way. Both are lists

**Comment by saeeddhqan** [comment]
some question:
does the Github API need a user cookie?
for searching users(repositories as well), the API is exactly like this URL: github.com/search?q=john&type=users
I mean, it shows description, email(if any), and other information?

**Comment by rachejazz** [comment]
Following information related to users:
```
      "avatar_url": "https://avatars.githubusercontent.com/u/42383989?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/rachejazz",
      "html_url": "https://github.com/rachejazz",
      "followers_url": "https://api.github.com/users/rachejazz/followers",
      "following_url": "https://api.github.com/users/rachejazz/following{/other_user}",
      "gists_url": "https://api.github.com/users/rachejazz/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/rachejazz/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/rachejazz/subscriptions",
      "organizations_url": "https://api.github.com/users/rachejazz/orgs",
      "repos_url": "https://api.github.com/users/rachejazz/repos",
      "events_url": "https://api.github.com/users/rachejazz/events{/privacy}",
      "received_events_url": "https://api.github.com/users/rachejazz/received_events",
 ```
With the starred ones, you can also view PRs made, merged, unmerged ones too

**Comment by rachejazz** [comment]
@saeeddhqan you haven't provided email in your account, so it does not show up here:
```
{
  "login": "saeeddhqan",
  "id": 31902891,
  "node_id": "MDQ6VXNlcjMxOTAyODkx",
  "avatar_url": "https://avatars.githubusercontent.com/u/31902891?v=4",
  "gravatar_id": "",
  "url": "https://api.github.com/users/saeeddhqan",
  "html_url": "https://github.com/saeeddhqan",
  "followers_url": "https://api.github.com/users/saeeddhqan/followers",
  "following_url": "https://api.github.com/users/saeeddhqan/following{/other_user}",
  "gists_url": "https://api.github.com/users/saeeddhqan/gists{/gist_id}",
  "starred_url": "https://api.github.com/users/saeeddhqan/starred{/owner}{/repo}",
  "subscriptions_url": "https://api.github.com/users/saeeddhqan/subscriptions",
  "organizations_url": "https://api.github.com/users/saeeddhqan/orgs",
  "repos_url": "https://api.github.com/users/saeeddhqan/repos",
  "events_url": "https://api.github.com/users/saeeddhqan/events{/privacy}",
  "received_events_url": "https://api.github.com/users/saeeddhqan/received_events",
  "type": "User",
  "site_admin": false,
  "name": "Saeed Dehqan",
  "company": "OWASP",
  "blog": "",
  "location": null,
  "email": null,
  "hireable": true,
  "bio": "OWASP Project Leader",
  "twitter_username": null,
  "public_repos": 5,
  "public_gists": 0,
  "followers": 57,
  "following": 0,
  "created_at": "2017-09-12T20:24:27Z",
  "updated_at": "2021-03-22T09:03:06Z"
}

```

**Comment by saeeddhqan** [comment]
I mean, it shows description, email(if any), and other information? If no, we need to add the API as a method like run_crawl_with_api() along with Github search without API: github.com/search

**Comment by rachejazz** [comment]
yes, it shows more information than the normal github search gives you

**Comment by rachejazz** [comment]
This is a subset of what the api can give you. Much much better than what a normal github search gives you:
[Follow this link](https://docs.github.com/en/rest/overview/endpoints-available-for-github-apps)

**Pull request merged**